### PR TITLE
Enable map stats in the DWRF flat map column writer

### PR DESCRIPTION
**Pull request by sdruzkin**
Enable map statistics in the DWRF flat map column writer. 

In this implementation map key stats use type specific statistics similar to Velox. 

Test plan:
- new unit tests
- validated in Validation Service

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/MapColumnStatisticsBuilder.java [hunk 1]
```diff
@@ -45,14 +53,22 @@ public void addMapStatistics(DwrfProto.KeyInfo key, ColumnStatistics columnStati
     {
         requireNonNull(key, "key is null");
         requireNonNull(columnStatistics, "columnStatistics is null");
-        nonNullValueCount += columnStatistics.getNumberOfValues();
```
Comment (sdruzkin) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/MapColumnStatisticsBuilder.java:48]: I made a mistake in the initial implementation confusing map row with map value. To fix it I added explicit increaseValueCount.

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/MapColumnStatisticsBuilder.java [hunk 2]
```diff
@@ -25,14 +25,22 @@
...
 public class MapColumnStatisticsBuilder
         implements StatisticsBuilder
 {
+    private final ImmutableList.Builder<MapStatisticsEntry> entries = new ImmutableList.Builder<>();
+    private final boolean collectKeyStats;
     private long nonNullValueCount;
     private boolean hasEntries;
-    private final ImmutableList.Builder<MapStatisticsEntry> entries = new ImmutableList.Builder<>();
+
+    public MapColumnStatisticsBuilder(boolean collectKeyStats)
```
Comment (sdruzkin) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/MapColumnStatisticsBuilder.java:40]: This new argument simplifies usage of the statistics builder for both cases with enabled and disabled map stats.

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/MapFlatColumnWriter.java [hunk 3]
```diff
@@ -228,15 +232,15 @@ public Map<Integer, ColumnStatistics> finishRowGroup()
         }
         Arrays.fill(valueWritersLastRow, 0);
 
-        ColumnStatistics mapStats = new ColumnStatistics((long) nonNullRowGroupValueCount, null);
-        rowGroupColumnStatistics.add(mapStats);
+        Map<Integer, ColumnStatistics> columnStatistics = getColumnStatisticsFromValueWriters(ColumnWriter::finishRowGroup, nonNullRowGroupValueCount);
```
Comment (sdruzkin) [inline presto-orc/src/main/java/com/facebook/presto/orc/writer/MapFlatColumnWriter.java:235]: Creation of stats for the map node moved to the getColumnStatisticsFromValueWriters.

File: presto-orc/src/test/java/com/facebook/presto/orc/CapturingOrcFileIntrospector.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
+/*
```
Comment (sdruzkin) [inline presto-orc/src/test/java/com/facebook/presto/orc/CapturingOrcFileIntrospector.java:1]: Extracted it as a top level class from the TestOrcFileIntrospector.

File: presto-orc/src/test/java/com/facebook/presto/orc/TestColumnStatistics.java [hunk 5]
```diff
@@ -0,0 +1,439 @@
...
+import static com.facebook.presto.orc.OrcEncoding.DWRF;
+import static com.facebook.presto.orc.OrcTester.createOrcWriter;
+import static com.facebook.presto.orc.OrcTester.mapType;
+import static com.facebook.presto.orc.metadata.CompressionKind.ZLIB;
+import static io.airlift.units.DataSize.Unit.MEGABYTE;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotNull;
+import static org.testng.Assert.assertNull;
+
+public class TestColumnStatistics
```
Comment (sdruzkin) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestColumnStatistics.java:61]: This is going to be a place for all column statistic tests enabling removal of OrcWriterValidation

File: presto-orc/src/test/java/com/facebook/presto/orc/TestColumnStatistics.java [hunk 6]
```diff
@@ -0,0 +1,439 @@
...
+            return KeyInfo.newBuilder().setBytesKey(ByteString.copyFromUtf8((String) key)).build();
+        }
+    }
+
+    private static void addIntStats(MapColumnStatisticsBuilder stats, Object key, long numberOfValues, long min, long max, long sum)
+    {
+        stats.addMapStatistics(keyInfo(key), new IntegerColumnStatistics(numberOfValues, null, new IntegerStatistics(min, max, sum)));
+    }
+
+    private static Map<Object, Object> mapOf(Object... values)
```
Comment (sdruzkin) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestColumnStatistics.java:431]: Created this to be able to create maps with null values.

File: presto-orc/src/test/java/com/facebook/presto/orc/metadata/statistics/TestMapColumnStatistics.java [hunk 7]
```diff
@@ -42,20 +42,23 @@ public Object[][] keySupplier()
     @Test(dataProvider = "keySupplier")
     public void testEqualsHashCode(KeyInfo[] keys)
     {
-        MapColumnStatisticsBuilder builder1 = new MapColumnStatisticsBuilder();
+        MapColumnStatisticsBuilder builder1 = new MapColumnStatisticsBuilder(true);
```
Comment (sdruzkin) [inline presto-orc/src/test/java/com/facebook/presto/orc/metadata/statistics/TestMapColumnStatistics.java:45]: Changed this test to follow up on my mistake in the initial implementation of MapColumnStatisticsBuilder

File: presto-orc/src/test/java/com/facebook/presto/orc/metadata/statistics/TestMapColumnStatisticsBuilder.java [hunk 8]
```diff
@@ -50,7 +50,7 @@ public Object[][] keySupplier()
     @Test
     public void testAddMapStatisticsNoValues()
     {
-        MapColumnStatisticsBuilder builder = new MapColumnStatisticsBuilder();
+        MapColumnStatisticsBuilder builder = new MapColumnStatisticsBuilder(true);
```
Comment (sdruzkin) [inline presto-orc/src/test/java/com/facebook/presto/orc/metadata/statistics/TestMapColumnStatisticsBuilder.java:53]: Changed this test to follow up on my mistake in the initial implementation of MapColumnStatisticsBuilder

File: presto-orc/src/test/java/com/facebook/presto/orc/metadata/TestDwrfMetadataReader.java [hunk 9]
```diff
@@ -355,17 +355,19 @@ public Object[][] columnStatisticsSupplier()
         ColumnStatistics integerColumnStatistics1 = new IntegerColumnStatistics(7L, null, new IntegerStatistics(25L, 95L, 100L));
         ColumnStatistics integerColumnStatistics2 = new IntegerColumnStatistics(9L, null, new IntegerStatistics(12L, 22L, 32L));
 
-        MapColumnStatisticsBuilder mapStatsIntKeyBuilder = new MapColumnStatisticsBuilder();
+        MapColumnStatisticsBuilder mapStatsIntKeyBuilder = new MapColumnStatisticsBuilder(true);
```
Comment (sdruzkin) [inline presto-orc/src/test/java/com/facebook/presto/orc/metadata/TestDwrfMetadataReader.java:358]: Changed this test to follow up on my mistake in the initial implementation of MapColumnStatisticsBuilder

**Review by arunthirupathi — APPROVED** [review]

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/MapColumnStatisticsBuilder.java [hunk 10]
```diff
@@ -86,16 +104,18 @@ public static Optional<MapStatistics> mergeMapStatistics(List<ColumnStatistics>
                     List<ColumnStatistics> allKeyStats = columnStatisticsByKey.computeIfAbsent(entry.getKey(), (k) -> new ArrayList<>());
                     allKeyStats.add(entry.getColumnStatistics());
                 }
+                nonNullValueCount += columnStatistics.getNumberOfValues();
             }
         }
 
         // merge all column stats for each key
-        MapColumnStatisticsBuilder mapStatisticsBuilder = new MapColumnStatisticsBuilder();
+        MapColumnStatisticsBuilder mapStatisticsBuilder = new MapColumnStatisticsBuilder(true);
```
[the remark this answers is not in the document]
Comment (sdruzkin) [reply presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/MapColumnStatisticsBuilder.java:112]: > Is this code path used for regular map

Regular MapColumnWriter doesn't use MapColumnStatisticsBuilder, that's why a code path for regular maps won't reach this place. However our team keep thinking about enabling map stats for regular maps that's why I'd prefer to keep the name as is.

-----

> We are setting the collectKeyStats to true unconditionally here.

That's correct and done intentionally. The idea here is that if the input has map entries, it means that the output should have map entries too. If the input doesn't have map entries, the method will produce regular ColumnStatistics.

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/MapColumnStatisticsBuilder.java [hunk 11]
```diff
@@ -25,14 +25,22 @@
...
 import static com.facebook.presto.orc.metadata.statistics.ColumnStatistics.mergeColumnStatistics;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
 import static java.util.Objects.requireNonNull;
 
 public class MapColumnStatisticsBuilder
         implements StatisticsBuilder
 {
+    private final ImmutableList.Builder<MapStatisticsEntry> entries = new ImmutableList.Builder<>();
+    private final boolean collectKeyStats;
```
[the remark this answers is not in the document]
Comment (sdruzkin) [reply presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/MapColumnStatisticsBuilder.java:36]: Done

**Comment by sdruzkin** [comment]
Tests are green. Good2merge

**Pull request merged**